cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -337,30 +337,93 @@ std::pair<VersionedItem, FrameAndDescriptor> LocalVersionedEngine::read_datafram
+        auto seg = std::move(std::get<1>(key_seg_pair));
+        auto meta = seg.has_metadata() ? std::make_optional<google::protobuf::Any>(*seg.metadata()) : std::nullopt;
+        auto start_index = seg.column(position_t(index::Fields::start_index)).type().visit_tag([&](auto column_desc_tag) -> NumericIndex {
+            using ColumnDescriptorType = std::decay_t<decltype(column_desc_tag)>;
+            using ColumnTagType =  typename ColumnDescriptorType::DataTypeTag;
+            if constexpr (is_numeric_type(ColumnTagType::data_type)) {
+                NumericIndex start_index = std::numeric_limits<NumericIndex>::max();
+                auto column_data = seg.column(position_t(index::Fields::start_index)).data();
+                while (auto block = column_data.template next<ColumnDescriptorType>()) {
+                    auto ptr = const_cast<NumericIndex *>(reinterpret_cast<const NumericIndex *>(block.value().data()));
+                    const auto row_count = block.value().row_count();
+                    for (auto i = 0u; i < row_count; ++i) {
+                        auto value = static_cast<NumericIndex>(*ptr++);
+                        start_index = value < start_index ? value : start_index;
+                    }
+                }
+                return start_index;
+            }else {
+                util::raise_rte("Unsoprted index type {}", seg.column(position_t(index::Fields::start_index)).type());
+            }
+        });
+
+        auto end_index = seg.column(position_t(index::Fields::end_index)).type().visit_tag([&](auto column_desc_tag) -> NumericIndex {
+            using ColumnDescriptorType = std::decay_t<decltype(column_desc_tag)>;
+            using ColumnTagType =  typename ColumnDescriptorType::DataTypeTag;
+            if constexpr (is_numeric_type(ColumnTagType::data_type)) {
+                NumericIndex end_index = 0;
+                auto column_data = seg.column(position_t(index::Fields::end_index)).data();
+                while (auto block = column_data.template next<ColumnDescriptorType>()) {
+                    auto ptr = const_cast<NumericIndex *>(reinterpret_cast<const NumericIndex *>(block.value().data()));
+                    const auto row_count = block.value().row_count();
+                    for (auto i = 0u; i < row_count; ++i) {
+                        auto value = static_cast<NumericIndex>(*ptr++);
+                        end_index = value > end_index ? value : end_index;
+                    }
+                }
+                return end_index;
+            }else {
+                util::raise_rte("Unsoprted index type {}", seg.column(position_t(index::Fields::end_index)).type());
+            }
+        });
+
+        auto end_index_opt = seg.row_count() == 0 ? std::nullopt : std::make_optional<IndexValue>(end_index);
+        auto start_index_opt = seg.row_count() == 0 ? std::nullopt : std::make_optional<IndexValue>(start_index);
```

Comment:
Have the lambdas above return optionals directly to avoid these two lines

Reply:
done